File: akk_example/src/main/java/com/hdu/chapter4_mailbox/MyPrioMailBox.java
package com.hdu.chapter4_mailbox;

import akka.actor.ActorSystem;
import akka.actor.PoisonPill;
import akka.dispatch.Envelope;
import akka.dispatch.PriorityGenerator;
import akka.dispatch.UnboundedStablePriorityMailbox;
import com.typesafe.config.Config;

import java.util.Comparator;

/**
 * DESCRIPTION:
 *
 * @author shushoufu
 * @Date 2018/6/20
 * @Time 下午7:08
 */
public class MyPrioMailBox extends UnboundedStablePriorityMailbox{
    public MyPrioMailBox(Comparator<Envelope> cmp, int initialCapacity) {
        super(cmp, initialCapacity);
    }

    public MyPrioMailBox(ActorSystem.Settings settings, Config config){
        super(new PriorityGenerator() {
            @Override
            public int gen(Object message) {
                if (message.equals("highpriority"))
                    return 0;
                else if (message.equals("lowpriority"))
                    return 2;
                else if (message.equals(PoisonPill.getInstance()))
                    return 3;
                else
                    return 1;
            }
        });
    }
}

File: akk_example/src/main/scala/com/hdu/mailbox/MyPrioMailBox.scala
package com.hdu.mailbox

import akka.actor.ActorSystem
import akka.dispatch.{PriorityGenerator, UnboundedStablePriorityMailbox}
import com.typesafe.config.Config

/**
 * DESCRIPTION:
 *
 * @author shushoufu
 * @Date 2018/4/11
 * @Time 下午8:36
 */
class MyPrioMailBox(setting: ActorSystem.Settings, config: Config) extends UnboundedStablePriorityMailbox(PriorityGenerator {
  // 'highpriority messages should be treated first if possible
  case 'highpriority ⇒ 0

  // 'lowpriority messages should be treated last if possible
  case 'lowpriority ⇒ 2

  // We default to 1, which is in between high and low
  case otherwise ⇒ 1
}) {

}
